### ISSUE TITLE
DDP loss计算疑问

### ISSUE THREAD

**Issue by wade0604** [Haochen-Wang409/U2PL#152]
您好！
请问在`loss = sup_loss + unsup_loss + contra_loss`中，为什么只对contra_loss进行dist.reduce操作，而sup_loss和unsup_loss都没有进行这个操作，我也看了一些其他DDP训练的loss计算方式，好像有不进行reduce操作直接backward的，也有先reduce再backward的，请问下这两种方式有区别吗？

**Comment by Haochen-Wang409** [comment]
contrastive loss 进行 reduce 是因为有可能某张卡采样到了 query，loss 正常计算；而另一张卡没有采样到 query，loss 直接跳过计算了。这样会导致不同卡之间的计算速度差距过大，有一定可能停止通讯。因此，我们让 contrastive loss 在各个卡之间统一。

其他的 loss 则没有这个问题，都会正常计算，因此无需人为地通讯。

**Comment by wade0604** [comment]
非常感谢您的解答！
